### PR TITLE
Expanding tests and docs relating to deleting profiles

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Improve docs and tests of commands relating to deleting profiles.

# Context

#8154 got me thinking about this stuff. 

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-collect-garbage.md:24]:
```suggestion
Instead, it looks looks in a few locations, and acts on all profiles it finds there:
```

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-collect-garbage.md:31]:
```suggestion
1. The default profile locations as specified in the [profiles] section of the manual.

2. `$NIX_STATE_DIR/profiles` and `$NIX_STATE_DIR/profiles/per-user`.

   With the exception of `$NIX_STATE_DIR/profiles/per-user/root` and `$NIX_STATE_DIR/profiles/default`, these directories are no longer used by other commands.
   `nix-collect-garbage` looks there anyways in order to clean up profiles from older versions of Nix.
```

The profiles page is not accessible on the manpages, but we could instead include it here and link and refer to that section.

File: doc/manual/src/command-ref/nix-env/delete-generations.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-env/delete-generations.md:39]:
```suggestion
The is because profiles are also garbage collection roots – any [store object] reachable from a profile is "alive" and ineligible for deletion.
```

File: doc/manual/src/command-ref/nix-env/delete-generations.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-env/delete-generations.md:59]:
```suggestion
Delete the generations numbered 3, 4, and 8, so long as the current active generation is not any of those.
```

File: doc/manual/src/command-ref/nix-env/delete-generations.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-env/delete-generations.md:70]:
```suggestion
Suppose `30` is the current generation, and we currently have generations numbered `20` through `32`.
```

I think this is brief enough to have it in a sentence.

**Review by fricklerhandwerk — APPROVED** [review]
This is clearly an improvement and should prevent lots of confusion over how things work.

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (thufschmitt) [inline doc/manual/src/command-ref/nix-collect-garbage.md:30]: I'm very much not keen on documenting that. These locations aren't something that people should care about and exposing them just increases our API surface for no good reason. People relying on `/nix/var/nix/profiles` is already what caused most of the troubles with https://github.com/NixOS/nix/pull/5226 (and similarly, https://github.com/NixOS/nix/pull/5588 is blocked by the fact that we expose more files than we should)

Comment (fricklerhandwerk) [reply doc/manual/src/command-ref/nix-collect-garbage.md:30]: We still have to answer the general question how to deal with such issues: There are some things that are accessible, i.e. are already part of the API, but should not be relied on.

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-collect-garbage.md:16]: I mentioned "variations" because the argument to that one is different depending on the flag/argument to this command.

Perhaps there is a better way to say this, like "running <cmd> with various arguments"?

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (fricklerhandwerk) [reply doc/manual/src/command-ref/nix-collect-garbage.md:19]: I see. This is a bit subtle and seems hard to express concisely. I'm not even sure if we should expand on "how it was done up til recently" in too much detail. That won't age well and just produce maintenance later on.

In this case, maybe something like "...in legacy locations".

Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-collect-garbage.md:19]: The global profiles (something I try to enshrine better on the next PR) will continue to be impacted, and they are intentionally shared between users, so I think the vague thing I originally wrote is still true even if we phase out the extra compat stuff?

**Comment by Ericson2314** [comment]
Thanks for the thorough review!

No comments on the test part though! Should we land that first? (I made it the first commit for a reason :D.)

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-collect-garbage.md:16]:
```suggestion
These options are the equivalent of running
[`nix-env --delete-generations`](@docroot@/command-ref/nix-env/delete-generations.md)
with various augments on multiple profiles,
prior to running `nix-collect-garbage` (or just `nix-store --gc`) without any flags.
```

File: doc/manual/src/command-ref/nix-env/delete-generations.md
Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-env/delete-generations.md:39]:
```suggestion
The is because profiles are also garbage collection roots — any [store object] reachable from a profile is "alive" and ineligible for deletion.
```

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-collect-garbage.md:26]:
```suggestion
These flags should be used with care, because they potentially delete generations of profiles used by other users on the system.
```

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-collect-garbage.md:58]: I did it below when I mentioned the command.

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-collect-garbage.md]: I added a note, but not marking this resolve yet until you both approve.

Comment (thufschmitt) [reply doc/manual/src/command-ref/nix-collect-garbage.md]: Looks good, thanks :+1:

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-collect-garbage.md:12]:
```suggestion
The command `nix-collect-garbage` is mostly an alias of [`nix-store --gc`](@docroot@/command-ref/nix-store/gc.md).
That is, it deletes all unreachable [store objects] in the Nix store to clean up your system.
```

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-collect-garbage.md:31]:
```suggestion
Instead, it looks in a few locations, and acts on all profiles it finds there:
```

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-collect-garbage.md:52]:
```suggestion
  Delete all old generations of profiles.
```

File: doc/manual/src/command-ref/nix-collect-garbage.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-collect-garbage.md:57]:
```suggestion
  Delete all generations of profiles older than the specified amount (except for the generations that were active at that point in time).
```

What does it mean to be active though? Which point in time exactly? Time of running the command, or falling into that interval between then and now?

File: doc/manual/src/command-ref/nix-env/delete-generations.md
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-env/delete-generations.md:31]: I really don't know what this is supposed to mean.

Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-env/delete-generations.md:31]: This is taking the parenthetical from above and making it a standalone sentence --- I am not sure exactly either!

We can figure it out and make it clear in a follow-up PR. For now, it is at least no worse than before.

Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-env/delete-generations.md:31]: > This command will delete all generations older than 30 days, except for the generation that was active 30 days ago (if it currently exists).

This is my interpretation

Comment (fricklerhandwerk) [reply doc/manual/src/command-ref/nix-env/delete-generations.md:31]: Maybe to slightly clarify then...

```suggestion
  The generation that was active at that point in the past is excluded, and will not be deleted.
```

Comment (Ericson2314) [reply doc/manual/src/command-ref/nix-env/delete-generations.md:31]: Yeah I like it. I will go read the code now to double check too.

**Pull request merged**
